File: src-backend/routers/quiz.py
"""Music quiz routes: session lifecycle, rounds, grading and suggestions."""
from fastapi import APIRouter, Body, HTTPException

import quiz

router = APIRouter()


def _session():
    try:
        return quiz.require_session()
    except quiz.QuizError as e:
        # 409 rather than 404: the route exists, there is just no game running
        # (e.g. the backend restarted mid-quiz), and the UI reacts by dropping
        # back to the setup screen.
        raise HTTPException(status_code=409, detail=str(e))


@router.post("/api/quiz/start")
def start_quiz(
    answer_style: str = Body("multiple_choice", embed=True),
    start_point: str = Body("beginning", embed=True),
    time_limit: int = Body(quiz.DEFAULT_TIME_LIMIT, embed=True),
):
    try:
        session = quiz.start_session(answer_style, start_point, time_limit)
        return {"settings": session.settings(), "round": quiz.next_round(session)}
    except quiz.QuizError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/api/quiz/next")
def next_quiz_round():
    session = _session()
    try:
        return {"round": quiz.next_round(session)}
    except quiz.QuizError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/api/quiz/answer")
def answer_quiz_round(
    track_id: str = Body("", embed=True),
    text: str = Body("", embed=True),
    elapsed: float = Body(0.0, embed=True),
):
    session = _session()
    try:
        return quiz.submit_answer(session, track_id=track_id, text=text, elapsed=elapsed)
    except quiz.QuizError as e:
        raise HTTPException(status_code=409, detail=str(e))


@router.get("/api/quiz/suggest")
def suggest_answers(q: str = "", limit: int = quiz.SUGGESTION_LIMIT):
    session = _session()
    return {"results": quiz.suggest(session, q, limit)}


@router.post("/api/quiz/stop")
def stop_quiz():
    """End the session and silence the snippet. Safe to call with none running,
    so the UI can fire it on navigating away without checking first."""
    quiz.end_session()
    return {"status": "ok"}

File: src/lib/utils/quiz.ts
import { apiUrl } from '$lib/backend';

export type QuizAnswerStyle = 'multiple_choice' | 'open_ended';
export type QuizStartPoint = 'beginning' | 'random';

export type QuizChoice = {
    id: string;
    title: string;
    artist_name: string;
    album_title: string;
    album_id: string;
};

export type QuizRound = {
    round_number: number;
    difficulty_level: number;
    time_limit: number;
    answer_style: QuizAnswerStyle;
    // Empty for open-ended rounds, where the player types instead of picking.
    choices: QuizChoice[];
    score: number;
    correct_count: number;
};

export type QuizResult = {
    correct: boolean;
    points: number;
    score: number;
    correct_count: number;
    round_number: number;
    answer: QuizChoice;
    selected_id: string;
};

export type QuizOptions = {
    answer_style: QuizAnswerStyle;
    start_point: QuizStartPoint;
    time_limit: number;
};

export const QUIZ_TIME_LIMITS = [5, 10, 15, 30];
export const QUIZ_DEFAULT_TIME_LIMIT = 10;

// The quiz session lives in the backend's memory only, so it can vanish under
// the page (a backend restart, a stop from another window). Distinguished from
// a real failure so the page can just drop back to the setup screen.
export class QuizSessionLost extends Error {}

async function post(path: string, body?: unknown): Promise<any> {
    const res = await fetch(apiUrl(path), {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body ?? {}),
    });
    if (res.status === 409) throw new QuizSessionLost('No quiz in progress');
    if (!res.ok) throw new Error(`Quiz request failed: ${path}`);
    return res.json();
}

export async function startQuiz(options: QuizOptions): Promise<QuizRound> {
    return (await post('/api/quiz/start', options)).round;
}

export async function nextRound(): Promise<QuizRound> {
    return (await post('/api/quiz/next')).round;
}

export async function submitAnswer(guess: {
    trackId?: string;
    text?: string;
    elapsed: number;
}): Promise<QuizResult> {
    return post('/api/quiz/answer', {
        track_id: guess.trackId ?? '',
        text: guess.text ?? '',
        elapsed: guess.elapsed,
    });
}

export async function stopQuiz(): Promise<void> {
    // Called on the way out of the page, where there is nothing left to
    // recover if it fails, and the backend treats "no session" as a no-op.
    try {
        await post('/api/quiz/stop');
    } catch {
        // ignored
    }
}

export async function fetchSuggestions(q: string): Promise<QuizChoice[]> {
    try {
        const res = await fetch(apiUrl(`/api/quiz/suggest?q=${encodeURIComponent(q)}`));
        if (!res.ok) return [];
        const data = await res.json();
        return Array.isArray(data.results) ? data.results : [];
    } catch {
        return [];
    }
}
